### ISSUE TITLE
Sorting of completion breaks fuzzy-matching

### ISSUE THREAD

**Issue by clarfonthey** [kakoune-lsp/kakoune-lsp#769]
Essentially: it appears that completion suggestions are sorted, rather than shown in the order they're given, which makes fuzzy-matching unable to provide precedence over "closer" matches.

Here's an example with `rust-analyzer`: `mem` suggests `core::arch::x86_64::_mm_cmp_epi8_max` among a ridiculous number of others before `std::mem`:

![described above](https://github.com/user-attachments/assets/f50d3106-b132-45ba-b870-f72c9d79487f)

**Comment by krobelus** [comment]
We sort completions according to the LSP `filterText` properties, and give them in that order to Kakoune, which preserves the orde if it's version v2023.08.05 or higher **and** if the "query string" (in this case `mem` is empty.
If the query string is non-empty, it is matched against all completion candidates; the match score dominates input order.
Only if two matches are considered equal score, the input order will be applied again.

With your example I don't get `std::mem` at all (rust-analyzer doesn't send it).
Perhaps the problem (if any) is on the rust-analyzer side.

**Comment by clarfonthey** [comment]
Huh, I was certain this was a `kak-lsp` bug, but I'll double-test in codium to make sure it is in fact a `rust-analyzer` bug. Maybe there's a subtle difference with the LSP implementation that they're relying on.

**Comment by krobelus** [comment]
So one issue is that while rust-analyzer sends `mem` when I type `me`, it no longer does when I type `mem`.
It would be useful because selecting that completion would auto-import `std::mem`, the same way it works when importing types or functions.

The rest of the issue comes down to Kakoune's fuzzy matching algorithm.
I saw that when typing `me`, `MetadataExt` was listed before `mem`.
I confirmed that this is the ranking assigned by Kakoune - the unit test `kak_assert(preferred("me", "MetadataExt", "mem"));` passes.
This is because word boundary matches are scored higher.
I think we can definitely improve the heuristic here; maybe rank a match higher if the majority of characters are matched (like in me/mem). Dunno
